src/gc.c

```diff
@@ -1827,28 +1890,31 @@ STATIC_INLINE jl_value_t *gc_mark_obj8(jl_ptls_t ptls, char *obj8_parent, uint8_
     jl_gc_markqueue_t *mq = &ptls->mark_queue;
     jl_value_t **slot = NULL;
     jl_value_t *new_obj = NULL;
+    int push_into_queue = gc_prequeue_free_space(mq) > 33;
     for (; obj8_begin < obj8_end; obj8_begin++) {
         slot = &((jl_value_t**)obj8_parent)[*obj8_begin];
         new_obj = *slot;
         if (new_obj != NULL) {
             verify_parent2("object", obj8_parent, slot, "field(%d)",
                             gc_slot_to_fieldidx(obj8_parent, slot, (jl_datatype_t*)jl_typeof(obj8_parent)));
-            gc_assert_parent_validity((jl_value_t *)obj8_parent, new_obj);
-            if (obj8_begin + 1 != obj8_end) {
-                gc_try_claim_and_push(mq, new_obj, &nptr);
-            }
-            else {
-                // Unroll marking of last item to avoid pushing
-                // and popping it right away
-                jl_taggedvalue_t *o = jl_astaggedvalue(new_obj);
-                nptr |= !gc_old(o->header);
-                if (!gc_try_setmark_tag(o, GC_MARKED)) new_obj = NULL;
+            assert(((uintptr_t)new_obj & 0x1) == 0);
+            if (push_into_queue) {
```

Comment:
I'm a bit confused about why this logic is not present in `mark_obj16` and `mark_obj32`.